### PR TITLE
Feature/scripts

### LINKED ISSUE #18: Script feature
Create the sequence of calls for a game.

### PR DESCRIPTION
**Pull request by vetom**
This closes #18.

### CODE REVIEW

**Comment by vetom** [comment]
Script diagram updated with new cards and a new section where moderator calls roles that he needs to know before.
nightlife

**Comment by vetom** [comment]
In the script, the asterisks symbol ( * ) means first night call only.

File: index.js [hunk 1]
```diff
@@ -24,6 +25,14 @@ exports.getAllCards = function () {
     return cards.getAll();
 }
 
+exports.getGameScript = function (game) {
```
Comment (wistcc) [inline index.js]: This function is not necessary. If I a have a game object I just need to pass `game.Deck` to the `getScriptFromDeck` function!

File: moderator.js [hunk 2]
```diff
@@ -0,0 +1,72 @@
+var lang = require('./languages');
+
+exports.getScriptFromGame = function (game) {
```
Comment (wistcc) [inline moderator.js]: This function is not necessary. If I a have a game object I just need to pass `game.Deck` to the `getScriptFromDeck` function!

**Review by wistcc — CHANGES_REQUESTED** [review]

File: moderator.js [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+
+exports.getScriptFromDeck = function (deck) {
+    return _getScriptFromDeck(deck);
+}
+
+function _getScriptFromDeck (deck) {
+    var script = '';
+
+    // To take note calls.
+    if(deck['Lycan']) script += ' - ' + lang.english['Lycan'] + '\n';
```
Comment (wistcc) [inline moderator.js]: [Single comment, not necessary to approve the PR] I think we should have another logic to do this, I suggest the next:
- Receive a string with the selected language.
- The `selectedLang` array should be sorted.
- Loop through the `selectedLang` array and do the check.

If we have a logic like this, we got some benefits:
- When we add other languages, we already have the logic to support them.
- DRY. We don't need to have so many `if` statements.
- DRY. We don't have to repeat this `' - '` and this `'\n'`.
- In order to add roles that are not in the deck, we could add some logic inside the loop to do this very easily.

File: moderator.js [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+    if(deck['Tough Guy']) script += ' - ' + lang.english['Tough Guy'] + '\n'; 
+    if(deck['Hunter']) script += ' - ' + lang.english['Hunter'] + '\n'; 
+    if(deck['Fruit Brute']) script += ' - ' + lang.english['Fruit Brute'] + '\n'; 
+    if(deck['Apprentice Seer']) script += ' - ' + lang.english['Apprentice Seer'] + '\n'; 
+    if(deck['Diseased']) script += ' - ' + lang.english['Diseased'] + '\n'; 
+    if(deck['Big Bad Wolf']) script += ' - ' + lang.english['Big Bad Wolf'] + '\n'; 
+
+    // Relationships
+    if(deck['Cupid']) {
+        script += ' - ' + lang.english['Cupid'] + '\n';
```
Comment (wistcc) [inline moderator.js]: We should merge these two because the are always called just once and one after the other.

**Review by wistcc — CHANGES_REQUESTED** [review]

File: index.js (hunk 1 above)
Comment (wovalle) [reply index.js]: Ir if it needed, we can do: `exports.getGameScript = moderator.getScriptFromDeck` without the invocation.

File: moderator.js (hunk 3 above)
Comment (wovalle) [reply moderator.js]: I agree with you Winner. But I would prefer to make an issue with this, internationalization is apart from scripting. 

Also, too many `script += ' - ' + whatever + '\n'`. I would do a function called `addToScript` (or something alike) and add the logic there.

File: languages.js [hunk 5]
```diff
@@ -0,0 +1,44 @@
+exports.english = {
```
Comment (vetom) [inline languages.js]: I will rename this to 'default'

Comment (wovalle) [reply languages.js]: I wouldn't rename it to default. This is the language pack in english. I would add that `logic` in the languageChooser (or whatever is named).

File: moderator.js (hunk 4 above)
Comment (wovalle) [reply moderator.js]: Agree.

**Review by wovalle — CHANGES_REQUESTED** [review]

File: languages.js (hunk 5 above)
Comment (vetom) [reply languages.js]: I thought to change it to default it's a good idea because we may have several English packs.

File: cards.js [hunk 6]
```diff
@@ -19,6 +19,9 @@ const cards  = [
     { role: 'Werewolf', value: -6, amount: 12 },
```
Comment (vetom) [inline cards.js]: I will order the cards alphabetically. I got some bad time verifying roles.

File: languages.js (hunk 5 above)
Comment (vetom) [reply languages.js]: Also, I will order these alphabetically.

File: moderator.js [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+    'Beholder', 
+    'The Count', 
+    'Insomniac',
+    'P.I.', 
+    'Old Hag', 
+    'Spellcaster', 
+    'Meeting'
+];
+
+exports.getScriptFromDeck = function (deck) {
```
Comment (wovalle) [inline moderator.js:5]: [Just a question] If you're exporting a private function, why not export the original function?

```javascript
exports.getScriptFromDeck = function (deck)  {
    var script = '';
    var calls = new Set();
    Object.keys(deck).forEach(function(role) {
        if(bindings[role]) bindings[role].forEach(function (call) {
            calls.add(call);
        });
    });
    secuence.forEach(function (call){
        if(calls.has(call)){
            script += '- ' + lang.default[call] + '\n';
        }
    });
    return script;
}
```

**Comment by wistcc** [comment]
@vetom Looks very good to me, just 2 things:
- Remember to add the `getTemplates` function.
- I would extract to other files the `bindings` and `sequence` objects

**Review by wistcc — APPROVED** [review]

**Comment by vetom** [comment]
@wistcc The templates function will be introduced in #28

File: moderator.js [hunk 8]
```diff
@@ -0,0 +1,23 @@
+var lang = require('./languages');
+var bindings = require('./bindings');
+var secuence = require('./secuence');
+
+exports.getScriptFromDeck = function (deck) {
```
Comment (wovalle) [inline moderator.js:5]: If you're exporting the private function, why not make it public from the beginning?

File: templates.js [hunk 9]
```diff
@@ -61,5 +65,23 @@ const templates = {
...
         'Drunk'
+    ],
+    vampires: [
+        'Vampire',
+        'Sorcerer',
+        'Cursed',
+        'Villager',
+        'Mayor',
+        'Prince',
+        'Tanner',
```
Comment (wovalle) [inline templates.js:76]: Tanner is a vampire role? wtf? :joy:

File: index.js [hunk 10]
```diff
@@ -23,6 +25,10 @@ exports.getAllCards = function () {
     return cards.getAll();
 }
 
+exports.getScriptFromDeck = function (deck) {
```
Comment (wovalle) [inline index.js:28]: If you're exporting the private function, why not make it public from the beginning?

**Review by wovalle — CHANGES_REQUESTED** [review]

File: templates.js (hunk 9 above)
Comment (vetom) [reply templates.js:76]: Not a vampire role! it's part of a vampires game 😂

File: index.js (hunk 10 above)
Comment (vetom) [reply index.js:28]: How can I do that❔

File: moderator.js (hunk 8 above)
Comment (wovalle) [reply moderator.js:5]:
```javascript
// Pass the reference 
exports.getScriptFromDeck = _getScriptFromDeck;

// Or simply
exports.getScriptFromDeck (deck) => {
     var script = '';
     var calls = new Set();
     Object.keys(deck).forEach(function(role) {
         if(bindings[role]) bindings[role].forEach(function (call) {
             calls.add(call);
         });
     });
     secuence.forEach(function (call){
         if(calls.has(call)){
             script += '- ' + lang.default[call] + '\n';
         }
     });
     return script;
 } 


```

**Review by wovalle — CHANGES_REQUESTED** [review]

**Review by wovalle — APPROVED** [review]

**Pull request merged**